mbed_devices/_internal/darwin/system_profiler.py

```diff
@@ -0,0 +1,51 @@
+"""Interactions with `system_profiler`."""
+import plistlib
+import re
+import subprocess
+from typing import List, Dict
+
+USBDeviceData = Dict
+
+
+def get_all_usb_devices_data() -> List[USBDeviceData]:
+    """Returns parsed output of `system_profiler` call."""
+    output = subprocess.check_output(["system_profiler", "-xml", "SPUSBDataType"], stderr=subprocess.DEVNULL)
+    if output:
+        data: List[USBDeviceData] = plistlib.loads(output)
+        return data
+    return []
+
+
+def get_end_usb_devices_data() -> List[USBDeviceData]:
+    """Returns only end devices from the output of `system_profiler` call."""
+    data = get_all_usb_devices_data()
+    return _filter_end_devices(data)
+
+
+def _filter_end_devices(data: List[USBDeviceData]) -> List[USBDeviceData]:
+    """Flattens the structure returned by `system_profiler` call.
+
+    Expected input will contain information about USB Buses and USB Hubs,
+    which we're not interested in. This function will return only "end devices",
+    relying on the fact they don't contain children under "_items" key.
+    """
+    end_devices = []
+    for device in data:
+        if "_items" in device:
+            child_devices = _filter_end_devices(device["_items"])
+            end_devices.extend(child_devices)
+        else:
+            end_devices.append(device)
+    return [device for device in end_devices if _is_end_device(device)]
+
+
+def _is_end_device(data: USBDeviceData) -> bool:
```

Comment:
Could you add a comment of what you mean by end device

Reply:
Fixed by writing more explicit code.